can/interfaces/ics_neovi/neovi_bus.py

```diff
@@ -0,0 +1,255 @@
+            return
+        for ics_msg in messages:
+            if ics_msg.NetworkID != self.network:
+                continue
+            if ics_msg.ArbIDOrHeader == 0:
+                # Looks like ICS device sends frames with ArbIDOrHeader = 0
+                # Need to find out exactly what they are for
+                # Filtering them for now
+                continue
+            if not self._is_filter_match(ics_msg.ArbIDOrHeader):
+                continue
+            self.rx_buffer.put(ics_msg)
+        if errors:
+            logger.warning("%d errors found" % errors)
+
+            for msg in ics.get_error_messages(self.dev):
+                logger.warning(msg)
+
+    def _is_filter_match(self, arb_id):
+        """
+        If SW filtering is used, checks if the `arb_id` matches any of
+        the filters setup.
+
+        :param int arb_id:
+            CAN ID to check against.
+
+        :return:
+            True if `arb_id` matches any filters
+            (or if SW filtering is not used).
+        """
+        if not self.sw_filters:
+            # Filtering done on HW or driver level or no filtering
+            return True
+        for can_filter in self.sw_filters:
+            if not (arb_id ^ can_filter['can_id']) & can_filter['can_mask']:
+                return True
+        return False
+
+    def _get_timestamp_for_msg(self, ics_msg):
+        if self._use_system_timestamp:
+            # This is the system time stamp.
+            # TimeSystem is loaded with the value received from the timeGetTime
+            # call in the WIN32 multimedia API.
+            #
+            # The timeGetTime accuracy is up to 1 millisecond. See the WIN32
+            # API documentation for more information.
+            #
+            # This timestamp is useful for time comparing with other system
+            # events or data which is not synced with the neoVI timestamp.
+            #
+            # Currently, TimeSystem2 is not used.
+            return ics_msg.TimeSystem
+        else:
+            # This is the hardware time stamp.
+            # The TimeStamp is reset to zero every time the OpenPort method is
+            # called.
+            return \
+                float(ics_msg.TimeHardware2) * self.ts_scaling[1] + \
+                float(ics_msg.TimeHardware) * self.ts_scaling[0]
+
+    def _ics_msg_to_message(self, ics_msg):
+        return Message(
+            timestamp=self._get_timestamp_for_msg(ics_msg),
+            arbitration_id=ics_msg.ArbIDOrHeader,
+            data=ics_msg.Data[:ics_msg.NumberBytesData],
+            dlc=ics_msg.NumberBytesData,
+            extended_id=bool(
+                ics_msg.StatusBitField & ics.SPY_STATUS_XTD_FRAME
+            ),
+            is_remote_frame=bool(
+                ics_msg.StatusBitField & ics.SPY_STATUS_REMOTE_FRAME
+            )
+        )
+
+    def recv(self, timeout=None):
+        try:
+            self._process_msg_queue(timeout=timeout)
```

Comment:
Shouldn't we check the rx_buffer for available messages first?

Reply:
Agree, however this is because the driver can only hold 20000 messages. So pumping the messages on each `recv` call is to avoid dropping messages. But, this is also downside to that.

See comment above regarding the Queue usage.